mlir/test/Dialect/Linalg/tile-tensors.mlir

```diff
@@ -167,3 +167,35 @@ module attributes {transform.with_named_sequence} {
     transform.yield
   }
 }
+
+// -----
+
+// CHECK-LABEL: func @non_monotonic_affine_expr
+//  CHECK-SAME:   %[[ARG0:[a-zA-Z0-9_]+]]: tensor<?xf32>
+func.func @non_monotonic_affine_expr(%arg0 : tensor<?xf32>) -> tensor<?xf32> {
+  %c0 = arith.constant 0 : index
+  %0 = tensor.dim %arg0, %c0 : tensor<?xf32>
+  %empty = tensor.empty(%0) : tensor<?xf32>
+
+  // CHECK: scf.for
+  // CHECK: %[[SIZE:[a-zA-Z0-9_]+]] = tensor.dim %[[ARG0]],
+  // CHECK: tensor.extract_slice %[[ARG0]][0] [%[[SIZE]]] [1] : tensor<?xf32> to tensor<?xf32>
+  %generic = linalg.generic
+    {indexing_maps = [affine_map<(d0) -> (d0 mod 3)>,
+                      affine_map<(d0) -> (d0)>],
+     iterator_types = ["parallel"]}
+    ins(%arg0: tensor<?xf32>)
+    outs(%empty : tensor<?xf32>) {
+    ^bb0(%in : f32, %out: f32):
+      linalg.yield %in : f32
+    } -> tensor<?xf32>
+  return %generic : tensor<?xf32>
+}
+
```

Comment:
How about adding one with static shapes like a variation of the example from the issue this solves:

```
#map1 = affine_map<(d0, d1) -> ((d0 + 1) mod 3, d1)>
#map2 = affine_map<(d0, d1) -> (d0, d1)>
module {
  func.func @test(%arg1: tensor<3x128xf32>) -> tensor<4x128xf32> {
    %1 = tensor.empty() : tensor<4x128xf32>
    %2 = linalg.generic {indexing_maps = [#map1, #map2], iterator_types = ["parallel", "parallel"]} 
        ins(%arg1 : tensor<3x128xf32>) outs(%1 : tensor<4x128xf32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<4x128xf32>
    return %2 : tensor<4x128xf32>
  }
}
```

This wraps around so the max would be found to be less than the min without this PR (and then I don't know how tiling would have behaved).

Reply:
Static and dynamic sizes are handled the same by the algorithm - its just that static shapes lead to constant propagation after the tile sizes have been computed.
So the wraparound behavior already happens on the dynamic test case.
I would say that the static test case makes sense once we modify the algorithm to make use of static information.

Reply:
Yes. I guess per the description this isn't supposed to be part of this PR?

_This is where we start to see the limits of the `AffineMap`s: these maps should have domain information, since they put in relation elements between affine sets that may be finite in the first place..._

Reply:
Dicussed offline.